src/Aspire.Dashboard/Utils/CompareHelpers.cs

```diff
@@ -0,0 +1,42 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Aspire.Dashboard.Utils;
+
+internal static class CompareHelpers
+{
+    // This method is used to compare two keys in a way that avoids timing attacks.
+    public static bool CompareKey(byte[] expectedKeyBytes, string requestKey)
+    {
+        const int StackAllocThreshold = 256;
+
+        var requestByteCount = Encoding.UTF8.GetByteCount(requestKey);
+
+        // A rented array could have previous data. However, we're trimming it to the exact byte count we need.
+        // That means all used bytes are overwritten by the following Encoding.GetBytes call.
+        byte[]? requestPooled = null;
+        var requestBytesSpan = (requestByteCount <= StackAllocThreshold ?
+            stackalloc byte[StackAllocThreshold] :
+            (requestPooled = ArrayPool<byte>.Shared.Rent(requestByteCount))).Slice(0, requestByteCount);
+
+        try
+        {
+            var encodedByteCount = Encoding.UTF8.GetBytes(requestKey, requestBytesSpan);
+            Debug.Assert(encodedByteCount == requestBytesSpan.Length, "Should match because span was previously trimmed to byte count value.");
+
+            return CryptographicOperations.FixedTimeEquals(expectedKeyBytes, requestBytesSpan);
+        }
+        finally
+        {
+            if (requestPooled != null)
+            {
+                ArrayPool<byte>.Shared.Return(requestPooled);
```

Comment:
Is the requestKey considered a secret? If so, we might want to clear it before returning to the pool.

Reply:
It is secret, but I'd like to leave it as-is. If someone has access to reading data in the array pool, then it's game over already. Do you feel strongly about this?

Reply:
It seems like bad practice. I've gotten similar feedback elsewhere: https://github.com/dotnet/aspire/pull/2740#discussion_r1518406490.

> If someone has access to reading data in the array pool, then it's game over already

Do we ever imagine allowing extensions in the dashboard? If we allow extensions, they can misuse stuff from the arraypool.

Reply:
The keys will also be IConfiguration. They're infinitely easier to get from there.

> Do we ever imagine allowing extensions in the dashboard? If we allow extensions, they can misuse stuff from the arraypool.

I doubt it, but who knows.